service-base/src/main/java/org/eclipse/hono/service/deviceconnection/AbstractDeviceConnectionAmqpEndpoint.java

```diff
@@ -0,0 +1,331 @@
+        final String tenantId = targetAddress.getTenantId();
+        final String deviceId = MessageHelper.getDeviceId(request);
+        final String gatewayId = MessageHelper.getGatewayId(request);
+
+        final Span span = TracingHelper.buildServerChildSpan(
+                tracer,
+                TracingHelper.extractSpanContext(tracer, request),
+                SPAN_NAME_SET_LAST_GATEWAY,
+                getClass().getSimpleName()
+        ).start();
+
+        final Future<Message> resultFuture;
+        if (tenantId == null || deviceId == null || gatewayId == null) {
+            TracingHelper.logError(span, "missing tenant, device and/or gateway");
+            resultFuture = Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST));
+        } else {
+            log.debug("setting last known gateway for tenant [{}], device [{}] to {}", tenantId, deviceId, gatewayId);
+
+            TracingHelper.TAG_TENANT_ID.set(span, tenantId);
+            TracingHelper.TAG_DEVICE_ID.set(span, deviceId);
+            TracingHelper.TAG_GATEWAY_ID.set(span, gatewayId);
+
+            resultFuture = getService().setLastKnownGatewayForDevice(tenantId, deviceId, gatewayId, span)
+                    .map(res -> DeviceConnectionConstants.getAmqpReply(
+                            DeviceConnectionConstants.DEVICE_CONNECTION_ENDPOINT,
+                            tenantId,
+                            request,
+                            res)
+                    );
+        }
+        return finishSpanOnFutureCompletion(span, resultFuture);
+    }
+
+    /**
+     * Processes a <em>get command handling protocol adapter instance</em> request message.
+     *
+     * @param request The request message.
+     * @param targetAddress The address the message is sent to.
+     * @return The response to send to the client via the event bus.
+     */
+    protected Future<Message> processGetCmdHandlingAdapterInstances(final Message request, final ResourceIdentifier targetAddress) {
+        final String tenantId = targetAddress.getTenantId();
+        final String deviceId = MessageHelper.getDeviceId(request);
+
+        JsonObject payload = null;
+        try {
+            payload = MessageHelper.getJsonPayload(request);
+        } catch (DecodeException e) {
+            logger.debug("failed to decode AMQP request message", e);
+            return Future.failedFuture(
+                    new ClientErrorException(
+                            HttpURLConnection.HTTP_BAD_REQUEST,
+                            "request message body contains malformed JSON"));
+        }
+
+        final Span span = TracingHelper.buildServerChildSpan(
+                tracer,
+                TracingHelper.extractSpanContext(tracer, request),
+                SPAN_NAME_GET_CMD_HANDLING_ADAPTER_INSTANCES,
+                getClass().getSimpleName()
+        ).start();
```

Comment:
How about creating the span before parsing the payload - then that error can be logged in the trace span.
See also https://github.com/eclipse/hono/commit/8f2958f6ba175f605ea13e431ba1efba8bed5f7f.

Reply:
@calohmn  Good catch. Fixed!